### ISSUE TITLE
difference of Decimal("0.67") vs Decimal(0.67)

### ISSUE THREAD

**Issue by nimo23** [ta4j/ta4j#86]
Is there a (rounding) difference of `Decimal("0.33")` and `Decimal(0.33)` or `Decimal.valueOf("-0.999")` and `Decimal.valueOf(-0.999)`? String will be converted directly to (Big)Decimal while the other uses double and convert it to (Big)Decimal. Any difference in precision?

**Comment by team172011** [comment]
After merging #84 i think both will have the same precision:

`BigDecimal(double val)` -> calls `BigDecimal.valueOf(double val)`
> Translates a double into a BigDecimal, using the double's canonical string representation provided by 
the Double.toString(double) method.
Note: This is generally the preferred way to convert a double (or float) into a BigDecimal, as the value returned is equal to that resulting from constructing a BigDecimal from the result of using Double.toString(double).

`Decimal(String val)` -> calls `BigDecimal(String val)` (above ends in the same with Double.toString(double))

But there is a big difference between `BigDecimal.valueOf(double val)` and `BigDecimal(double val)`

`public BigDecimal(double val)`
> Translates a double into a BigDecimal which is the exact decimal representation of the double's binary floating-point value. The scale of the returned BigDecimal is the smallest value such that (10scale × val) is an integer.

> Notes:
The results of this constructor can be somewhat unpredictable. One might assume that writing new BigDecimal(0.1) in Java creates a BigDecimal which is exactly equal to 0.1 (an unscaled value of 1, with a scale of 1), but it is actually equal to 0.1000000000000000055511151231257827021181583404541015625. This is because 0.1 cannot be represented exactly as a double (or, for that matter, as a binary fraction of any finite length). Thus, the value that is being passed in to the constructor is not exactly equal to 0.1, appearances notwithstanding.
    The String constructor, on the other hand, is perfectly predictable: writing new BigDecimal("0.1") creates a BigDecimal which is exactly equal to 0.1, as one would expect. Therefore, it is generally recommended that the String constructor be used in preference to this one.
    When a double must be used as a source for a BigDecimal, note that this constructor provides an exact conversion; it does not give the same result as converting the double to a String using the Double.toString(double) method and then using the BigDecimal(String) constructor. To get that result, use the static valueOf(double) method.

**Comment by nimo23** [comment]
Simple question:

**Will `Decimal.valueOf("-0.999") `and `Decimal.valueOf(-0.999)` will result in the same precision?**

**Yes or no?**

Acctually, `Decimal.valueOf(double val)` convert double to `(Big)Decima`l (which can result in floating point issues), where `Decimal.valueOf(String val)` convert the string value directly to `(Big)Decimal` - the string conversion (is slower) but leads to a higher precision than converting double to BigDecimal! 

User does not know about this implications when using Decimal.valueOf(double val), hence the Decimal.class should convert the provided double to string and use the string as a input for converting to BigDecimal. This ensures that precision is not lost. Because when  Decimal.valueOf(double val) elsewhere within ta4j can lead to lower precision, we have both of the bad sides: low performance (because of BigDecimal) and low precision (because of the conversion process).

Please check this.

**Comment by team172011** [comment]
I don't know if it makes a difference **of interest** at the precision regarding the last decimal place of the initial double value:

```
    @Test
    public void decimalBehaviorTest(){
        double d = 2.000000312;
        Decimal decimalValueOfd = Decimal.valueOf(d);
        Decimal decimalValueOfs = Decimal.valueOf("2.000000312");

        System.out.println(decimalValueOfd.multipliedBy(decimalValueOfd));
        System.out.println(decimalValueOfs.multipliedBy(decimalValueOfd));
        System.out.println(d*d);
    }
```

```
4.0000012480000975752863577309254
4.0000012480000974596431788654627
4.000001248000098
```

After calculating with the `BigDecimal` instances the behavior of those decimal places is different, but booth are not like the user would expect, or not?

**Comment by nimo23** [comment]
So we can say, this:

Decimal decimalValueOfs = Decimal.valueOf("2.000000312");

is more precice than the other two versions. 

Thus, the "Decimal.class" or its new implementation ("ExactDecimal") should convert the provided double to string and use the string as a input for converting to BigDecimal. This is the slowest procedure, but the "exact".

As described above:

> The String constructor, on the other hand, is perfectly predictable: writing new BigDecimal("0.1") creates a BigDecimal which is exactly equal to 0.1, as one would expect. Therefore, it is generally recommended that the String constructor be used in preference to this one.

**Comment by katsadim** [comment]
@team172011 this is an important case that you make by providing the unit test. I can see two issues here:
1.  `Decimal::multipliedBy` violates symmetricity (that means that `x.multipliedBy(y)` does not yield the same result as `y.multipliedBy(x)` )
2.  After a BigDecimal multiplication, the number of decimal places is increased, as opposed to the double  multiplication where the number of decimal places is limited

The first issue is fixed by my recent PR #84. 
For the second issue: I think BigDecimal's behaviour is the desired behaviour after a calculation. My [comment](https://github.com/ta4j/ta4j/pull/84#issuecomment-344045756) on the PR was referring only to Decimal initialisation where we don't care about the extra decimal places added by the BigDecimal class. But in the case of a calculation we need to preserve as many decimal places as we can so as to keep precision at acceptable levels after repeated calculations.

@nimo23 This is `BigDecimal::valueOf` implementation:
```java
public static BigDecimal valueOf(double val) {
        return new BigDecimal(Double.toString(val));
    }
``` 
It internally converts the double to a String therefore there is no need for `ExactDecimal` to accept only String arguments. So we are covered by PR #84.

**Comment by team172011** [comment]
that answers @nimo23 question and I have merged PR to `develop`
